designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,505 @@
+        ignores: ["**/*.test.js"],
+        languageOptions: {
+            sourceType: "commonjs",
+            globals: {
+                ...globals.node
+            }
+        }
+    },
+
+    // intersected files and merged ignores
+    {
+        name: "myconfig > config2",
+        files: [["**/src/*.js", "**/*.js"]],
+        ignores: ["**/*.test.js", "**/tests"],
+        rules: {
+            "no-console": "error"
+        }
+    },
+
+    // original config
+    {
+        name: "myconfig",
+        files: ["**/src/*.js"],
+        ignores: ["**/*.test.js"]
+        rules: {
+            semi: "error"
+        }
+    }
+
+];
+```
+
+Notes:
+
+1. The `files` and `ignores` values from the base config always come first in the calculated `files` and `ignores`. If there's not a match, then there's no point in continuing on to use the patterns from the extended configs.
+1. The `name` key is generated for calculated config objects so that it indicates the inheritance from another config. The extended configs don't exist in the final representation of the config array.
+1. These behaviors differ from the [typescript-eslint extends helper](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/typescript-eslint/src/config-helper.ts).
+
+#### Extending Arrays
+
+Arrays can also be used in `extends` (to eliminate the guesswork of what type a config is). When evaluating `extends`, ESLint internally calls `.flat(Infinity)` on the array and then processes the config objects as discussed in the previous example. Consider the following:
+
+```js
+import js from "@eslint/js";
+import example from "eslint-plugin-example";
+
+export default [
+
+    {
+        files: ["**/src/*.js"],
+        extends: [
+            [js.configs.recommended, example.configs.recommended]
+        ]
+    }
+
+];
+```
+
+This is equivalent to the following:
+
+```js
+import js from "@eslint/js";
+import example from "eslint-plugin-example";
+
+export default [
+
+    {
+        files: ["**/src/*.js"],
+        extends: [js.configs.recommended, example.configs.recommended]
+    }
+
+];
+```
+
+The extended objects are evaluated in the same order and result in the same final config.
+
+#### Extending Named Configs
+
+While the previous examples are an improvement over current situation, we can go a step further and restore the use of named configs by allowing strings inside of the `extends` array. When provided, the string must refer to a config contained in a plugin. For example:
+
+```js
+import js from "@eslint/js";
+
+export default [
+
+    {
+        plugins: {
+            js
+        },
+        files: ["**/src/*.js"],
+        extends: ["js/recommended"]
+    }
+
+];
+```
+
+Here, `js/recommended` refers to the plugin defined as `js`. Internally, ESLint will look up the plugin with the name `js`, looks at the `configs` key, and retrieve the `recommended` key as a replacement for the string `"js/recommended"`. 
```

Comment:
I'm not sure I understand what you're saying. This proposal would solve the problem you're highlighting here.

Reply:
@nzakas This proposal contains many proposals, one of which aims to solve this by adding magic that modifies the config. That part of the proposal feels like a suggestion that should be made separately as a follow up to this RFC and is needlessly complicated compared to what I highlight / suggest here.

Reply:
I don't think using the term "magic" helpful in this context. I'm using the same lookup pattern that we already use for rules, processors, and languages, so this is actually a well-established pattern. Further, the implementation of `extends` may change depending on whether or not this part of the proposal is included, so it makes sense to include it in this RFC.

What you're suggesting is that everyone who writes plugins change the way they do it AND we change the way we interpret plugins. That's a much more dramatic proposal than named configs.

Reply:
What I’m suggesting is not drastic at all. Not sure why you are suggesting that it is?

I’m simply suggesting that ESLint does not try to enforce a standard for configs exported by plugins – keeping configs completely separate from the plugin object.

Also: I’m sorry that “magic” offended you, I should have used “logic” as a more neutral word. I intended “magic” to be interpreted as “resolving and inserting plugin reference automagically”. I’ll try to be more careful with my choice of words. Thanks for pointing it out 🙏

Reply:
> I’m simply suggesting that ESLint does not try to enforce a standard for configs exported by plugins – keeping configs completely separate from the plugin object.

Yes, I understand what you were suggesting. However, the team has already agreed to [standardize config exports](https://github.com/eslint/eslint/issues/18095), so encouraging plugin developers to use the already-defined plugin format is part of the goal of this RFC.

Reply:
This RFC should really refer to such a crucial change request then as its essential background information for anyone who it has passed by.

Its also kind of confusing what becomes RFC:s and what doesn't. That change request would have been great to have as an RFC so that it would have eg appeared on my radar and not mostly/just on the radar of those following all issues in the ESLint repository. Or is there something I'm missing where I should have made myself aware of this?

As I'm a co-maintainer of quite a few plugins I have made sure to follow the RFC:s, but I do not have the bandwidth to follow all of the ESLint issues and I'm not a member of the ESLint team so I shouldn't have to be following it to give my feedback as a community member?

Reply:
I'm sorry, I don't know how to solve the problem of making sure everyone is aware of any changes they're interested in at any point in time. The biggest changes are always done via RFC, and we promote those on Twitter and Mastodon for maximum visibility.

Standardizing something, or indeed, deciding to document something that was previously undocumented, doesn't really lend itself to the RFC process.

In this case, we're really talking about a documentation change for what we'd like to see people do. We have no way to enforce such a convention and this documentation change won't affect the way existing packages can be used.